netutils/acl.py

```diff
@@ -179,55 +207,80 @@ def __init__(self, data: t.Any, *args: t.Any, **kwargs: t.Any):  # pylint: disab
         Examples:
             >>> from netutils.acl import ACLRule
             >>>
-            >>> acl_data = dict(
+            >>> rule = ACLRule(
             ...     name="Check no match",
             ...     src_ip=["10.1.1.1"],
             ...     dst_ip="172.16.0.10",
             ...     dst_port="tcp/www-http",
             ...     action="permit",
             ... )
             >>>
-            >>> rule = ACLRule(acl_data)
             >>>
-            >>> rule.expanded_rules
+            >>> rule._expanded_rules
             [{'name': 'Check no match', 'src_ip': '10.1.1.1', 'dst_ip': '172.16.0.10', 'dst_port': '6/80', 'action': 'permit'}]
             >>>
         """
-        self.processed: t.Dict[str, str] = {}
-        self.data = data
-        self.load_data()
+        self.__load_data(kwargs=kwargs)
 
-    def load_data(self) -> None:
+    def __load_data(self, kwargs) -> None:
         """Load the data into the rule while verifying input data, result data, and processing data."""
+        # Remaining kwargs stored under ACLRule.Meta
+        pop_kwargs = []
+        for key, val in kwargs.items():
+            if key not in get_attributes(self):
+                setattr(self.Meta, key, val)
+                pop_kwargs.append(key)
+
+        # Pop unneeded keys
+        for key in pop_kwargs:
+            kwargs.pop(key)
+
+        # Ensure each class attr is in init kwargs.
+        for attr in get_attributes(self):
+            if attr not in kwargs:
+                kwargs[attr] = getattr(self, attr)
+
+        # Store the init input
+        self._preprocessed_data = copy.deepcopy(kwargs)
+        self._processed_data = copy.deepcopy(self._preprocessed_data)
+
+        # Input check
         self.input_data_check()
-        for attr in self.attrs:
-            if not self.data.get(attr):
-                continue
-            if hasattr(self, f"process_{attr}"):
-                proccessor = getattr(self, f"process_{attr}")
-                _attr_data = proccessor(self.data[attr])
+
+        for attr in get_attributes(self):
+            processor_func = getattr(self, f"process_{attr}", None)
+            if processor_func:
+                _attr_data = processor_func(self._processed_data[attr])
             else:
-                _attr_data = self.data[attr]
-            self.processed[attr] = _attr_data
+                _attr_data = self._processed_data[attr]
+
+            self._processed_data[attr] = _attr_data
             setattr(self, attr, _attr_data)
+
         self.result_data_check()
         self.validate()
-        self.expanded_rules = _cartesian_product(self.processed)
-        if self.filter_same_ip:
-            self.expanded_rules = [item for item in self.expanded_rules if item["dst_ip"] != item["src_ip"]]
+
+    @property
+    def _expanded_rules(self):
```

Comment:
Why make this private? I think there is value in having access to this and perhaps in serialize format as well.

Reply:
Agreed